# Stabilized SU(2) reblocking for the S4 indicator proof

This notebook checks the finite symmetry-data calculation in
`JOP_revision/manuscript.tex`, Section `sec:su2_reduction`, and the explicit
example in Appendix `app:SU2_example`.

The manuscript's main integer problem is

`S @ u = b + EBR @ v`, with `u,v >= 0` integral,

where `b` runs over all admissible two-band vectors, `S` contains only the
16 determinant-one two-band generators, and `EBR` contains the 18
double-valued atomic columns. No scalar generators enter this main check.
Two supplemental checks also consider scalar U(1) generators and one-band targets.

At an S4-fixed momentum a gauge transformation conjugates the sewing matrix,
so its determinant cannot change. The direct-reduction criterion is
`det B(K)=1` at Gamma, Z, M, A and `det D(X)=1`; compatibility propagates
the latter to R. These constraints select the columns of `S`.

Scope: the notebook verifies nonnegative integer symmetry-data identities.
It does not construct smooth sewing matrices, perform the global gauge
transformation, or prove that equality of symmetry data implies a smooth
reblocking of the occupied bundle. Those require separate analytical arguments.

In [1]:
from __future__ import annotations

import re
import sys
import time
from itertools import product
from pathlib import Path

import numpy as np


def find_data_dir() -> Path:
    """Find essay_script whether the notebook is launched there or from the repo root."""
    cwd = Path.cwd().resolve()
    candidates = [cwd, cwd / "essay_script", cwd.parent / "essay_script"]
    for candidate in candidates:
        if (candidate / "ebr_raw_data8133.txt").exists() and (candidate / "gu_cover_solver.py").exists():
            return candidate
    raise FileNotFoundError("Could not find essay_script data files from current working directory")


DATA_DIR = find_data_dir()
sys.path.insert(0, str(DATA_DIR))

N_COLS = 36
FULL_N_ROWS = 40
N_ROWS = 20
ROW_OFFSET = {"Gamma": 0, "A": 8, "M": 16, "Z": 24, "R": 32, "X": 36}
KPOINT_TYPES = ("Gamma", "A", "M", "Z", "R", "X")
OLD_ROWS = (
    list(range(4, 8))
    + list(range(12, 16))
    + list(range(20, 24))
    + list(range(28, 32))
    + [34, 35]
    + [38, 39]
)

BLOCKS = [
    ("Gamma", [0, 1, 2, 3]),
    ("A", [4, 5, 6, 7]),
    ("M", [8, 9, 10, 11]),
    ("Z", [12, 13, 14, 15]),
    ("R", [16, 17]),
    ("X", [18, 19]),
]
S4_BLOCKS = BLOCKS[:4]
C2_BLOCKS = BLOCKS[4:]

row_labels = (
    [f"Gamma_{i}" for i in range(5, 9)]
    + [f"A_{i}" for i in range(5, 9)]
    + [f"M_{i}" for i in range(5, 9)]
    + [f"Z_{i}" for i in range(5, 9)]
    + [f"R_{i}" for i in range(3, 5)]
    + [f"X_{i}" for i in range(3, 5)]
)


def parse_cell(cell_html: str) -> list[int]:
    """Extract representation sub-indices from one BCS table cell."""
    parts = re.split(r"&nbsp;&oplus;&nbsp;|&oplus;", cell_html)
    indices = []
    for part in parts:
        match = re.search(r"<sub>(\d+)</sub>", part)
        if match:
            indices.append(int(match.group(1)))
    return indices


def _row_label(row_html: str) -> str:
    cells = re.findall(r"<td[^>]*>(.*?)</td>", row_html, re.DOTALL)
    if not cells:
        return ""
    return re.sub(r"<[^>]+>", "", cells[0]).strip()


def _identify_kpoint(label: str) -> str | None:
    label = label.replace("?", "Gamma").replace("&Gamma;", "Gamma")
    for rep in KPOINT_TYPES:
        if re.match(rf"^{rep}\s*:", label):
            return rep
    return None


def _extract_table_rows(html: str) -> list[str]:
    # The copied BCS HTML may omit closing </tr> tags, so split on opening tags.
    return [part for part in re.split(r"<tr>", html, flags=re.IGNORECASE) if "<td" in part]


def build_B_from_html(html: str) -> np.ndarray:
    data_rows: list[tuple[str, str]] = []
    for row_html in _extract_table_rows(html):
        rep_type = _identify_kpoint(_row_label(row_html))
        if rep_type is not None:
            data_rows.append((rep_type, row_html))

    if len(data_rows) != 6:
        labels = [_row_label(row) for _, row in data_rows]
        raise ValueError(f"Expected 6 k-point rows, found {len(data_rows)}: {labels}")

    first_cells = re.findall(r"<td[^>]*>(.*?)</td>", data_rows[0][1], re.DOTALL)
    n_label_cols = len(first_cells) - N_COLS
    if n_label_cols < 1:
        raise ValueError(f"Expected {N_COLS} data columns, got {len(first_cells)} cells")

    B = np.zeros((FULL_N_ROWS, N_COLS), dtype=int)
    for col_idx in range(N_COLS):
        for rep_type, row_html in data_rows:
            cells = re.findall(r"<td[^>]*>(.*?)</td>", row_html, re.DOTALL)
            cell_html = cells[n_label_cols + col_idx]
            for sub_idx in parse_cell(cell_html):
                B[ROW_OFFSET[rep_type] + sub_idx - 1, col_idx] += 1
    return B


raw_path = DATA_DIR / "ebr_raw_data8133.txt"
EBR_full = build_B_from_html(raw_path.read_text(encoding="utf-8"))
EBR_projected = EBR_full[OLD_ROWS, :]
# Retain only the 18 double-valued EBRs, without permuting R/X rows.
ebr_source_columns = np.flatnonzero(np.any(EBR_projected != 0, axis=0))
EBR = EBR_projected[:, ebr_source_columns].copy()
single_valued_rows = np.setdiff1d(np.arange(FULL_N_ROWS), OLD_ROWS)
assert EBR.shape == (20, 18)
assert not np.any(EBR_full[np.ix_(single_valued_rows, ebr_source_columns)])

# Compatibility matrix in the manuscript symmetry-data row order.
CR = np.zeros((8, N_ROWS), dtype=int)
CR[0, 0] = CR[0, 1] = 1
CR[0, 12] = CR[0, 13] = -1
CR[1, 2] = CR[1, 3] = 1
CR[1, 14] = CR[1, 15] = -1
CR[2, 4] = CR[2, 5] = 1
CR[2, 8] = CR[2, 9] = -1
CR[3, 6] = CR[3, 7] = 1
CR[3, 10] = CR[3, 11] = -1
CR[4, 16] = 1
CR[4, 18] = -1
CR[5, 17] = 1
CR[5, 19] = -1
CR[6, 0:4] = 1
CR[6, 16] = CR[6, 17] = -1
CR[7, 4:8] = 1
CR[7, 16] = CR[7, 17] = -1

# Manuscript convention: R_3 / R_4 have C2 eigenvalues -i / +i,
# so n_R^(1/2) = m(R_3), n_R^(-1/2) = m(R_4).
# z_4S contains +n_R^(1/2) - n_R^(-1/2): unscaled R entries (+1,-1).
# P_1 = 2*p_1, hence z_4S=0 means P_1 @ B = 0 mod 8.
P_1 = np.array([0, 0, 0, 0, 1.5, -0.5, -1.5, 0.5, 0, 0, 0, 0, 1.5, -0.5, -1.5, 0.5, 1.0, -1.0, 0, 0])
P_2 = np.array([-1, 0, 1, 0, 1, 0, -1, 0, -1, 0, 1, 0, 1, 0, -1, 0, 0, 0, 0, 0], dtype=int)
P_1 = (2 * P_1).astype(int)

assert np.all(CR @ EBR == 0)
assert np.all(P_1 @ EBR % 8 == 0), "Atomic EBRs must have z_4S=0"
assert np.all(P_2 @ EBR % 2 == 0), "Atomic EBRs must have delta_2S=0"

print("data directory:", DATA_DIR)
print("EBR shape:", EBR.shape)
print("original MBANDREP columns (1-based):", (ebr_source_columns + 1).tolist())
print("P_1 R coefficients:", P_1[16:18].tolist())
print("all EBR indicator congruences vanish:", bool(np.all(P_1 @ EBR % 8 == 0) and np.all(P_2 @ EBR % 2 == 0)))
print("CR @ EBR == 0:", bool(np.all(CR @ EBR == 0)))
print("row labels:", row_labels)

data directory: D:\yao\PkU\Song Group\Symmetry Representation and Topological Invariant\essay_script
EBR shape: (20, 18)
original MBANDREP columns (1-based): [5, 6, 7, 8, 13, 14, 15, 16, 21, 22, 23, 24, 29, 30, 31, 32, 35, 36]
P_1 R coefficients: [2, -2]
all EBR indicator congruences vanish: True
CR @ EBR == 0: True
row labels: ['Gamma_5', 'Gamma_6', 'Gamma_7', 'Gamma_8', 'A_5', 'A_6', 'A_7', 'A_8', 'M_5', 'M_6', 'M_7', 'M_8', 'Z_5', 'Z_6', 'Z_7', 'Z_8', 'R_3', 'R_4', 'X_3', 'X_4']


## Enumerate one-band and two-band blocks

The targets are exactly the 1-band and 2-band symmetry-data vectors satisfying compatibility plus `z_4S=delta_2S=0`, following the manuscript definition of the monoid M.

For the two-band `SU(2)` criterion, the determinant obstruction must vanish at both the `S4`- and `C2`-invariant momenta. In each `S4`-fixed block `(K_5,K_6,K_7,K_8)` at `Gamma,Z,M,A`, the conjugate pairs are `(K_5,K_7)` and `(K_6,K_8)`, so these pair multiplicities must match. At `R` and `X`, the conjugate `C2` pair multiplicities `(K_3,K_4)` must also match, so `det D(R)=det D(X)=1`. No one-band block is classified as `SU(1)` below, since its fixed-point eigenvalues are not all equal to `1`.

In [2]:
def compositions(total: int, parts: int):
    """Non-negative integer tuples of length parts summing to total."""
    if parts == 1:
        yield (total,)
        return
    for first in range(total + 1):
        for rest in compositions(total - first, parts - 1):
            yield (first,) + rest


def enumerate_band_data(N: int, mods=(8, 2)) -> np.ndarray:
    """All N-band B >= 0 satisfying compatibility, z_4S=0, and delta_2S=0."""
    block_choices = [list(compositions(N, len(idx))) for _, idx in BLOCKS]
    results = []
    for combo in product(*block_choices):
        B = np.zeros(N_ROWS, dtype=int)
        for (_, idx), vals in zip(BLOCKS, combo):
            B[idx] = vals
        if np.any(CR @ B != 0):
            continue
        if (P_1 @ B) % mods[0] != 0:
            continue
        if (P_2 @ B) % mods[1] != 0:
            continue
        results.append(B)
    return np.array(results, dtype=int).reshape(-1, N_ROWS)


def z2_numerator(B: np.ndarray) -> int:
    B = np.asarray(B, dtype=int).reshape(-1)
    return int(B[1] + B[5] + B[9] + B[13] - (B[0] + B[4] + B[8] + B[12]))


def z2(B: np.ndarray) -> int:
    num = z2_numerator(B)
    if num % 2 != 0:
        raise ValueError(f"z2 numerator is odd for {B}: {num}")
    return (num // 2) % 2


def is_su2_candidate(B: np.ndarray) -> bool:
    """Two-band data with det B(K)=1 at Gamma,Z,M,A and det D(K)=1 at R,X."""
    B = np.asarray(B, dtype=int).reshape(-1)
    if any(int(B[idx].sum()) != 2 for _, idx in BLOCKS):
        return False
    for _, idx in S4_BLOCKS:
        if B[idx[0]] != B[idx[2]]:  # K_5 paired with K_7
            return False
        if B[idx[1]] != B[idx[3]]:  # K_6 paired with K_8
            return False
    for _, idx in C2_BLOCKS:
        if B[idx[0]] != B[idx[1]]:  # K_3 paired with K_4
            return False
    return True


def has_equal_c2_pair_somewhere(B: np.ndarray) -> bool:
    """Diagnostic used in the TeX: both bands have the same C2 eigenvalue at some S4-fixed K."""
    B = np.asarray(B, dtype=int).reshape(-1)
    for _, idx in S4_BLOCKS:
        if B[idx[0]] + B[idx[1]] == 2:
            return True
        if B[idx[2]] + B[idx[3]] == 2:
            return True
    return False


def active_rep_labels(B: np.ndarray) -> list[str]:
    B = np.asarray(B, dtype=int).reshape(-1)
    return [f"{row_labels[i]}={B[i]}" for i in np.nonzero(B)[0]]


B1 = enumerate_band_data(1)
B2 = enumerate_band_data(2)
su2_mask = np.array([is_su2_candidate(b) for b in B2], dtype=bool)
B2_su2 = B2[su2_mask]
su2_block_ids = np.where(su2_mask)[0]

assert len(B1) == 16 and len(B2) == 438 and len(B2_su2) == 16
assert all(z2(b) == 0 for b in B1)
assert all(z2(e) == 0 for e in EBR.T)

print("B1 count:", len(B1))
print("B2 count:", len(B2))
print("B1 z2 values:", sorted({z2(b) for b in B1}))
print("B2 z2 counts:", {value: int(sum(z2(b) == value for b in B2)) for value in (0, 1)})
print("SU(2)-candidate B2 count:", len(B2_su2))
print("SU(2)-candidate B2 z2 counts:", {value: int(sum(is_su2_candidate(b) and z2(b) == value for b in B2)) for value in (0, 1)})
print("z2=1 B2 blocks not directly SU(2)-reducible:", int(sum((not is_su2_candidate(b)) and z2(b) == 1 for b in B2)))
print("z2=1 B2 blocks with equal C2 pair at some fixed K:", int(sum(has_equal_c2_pair_somewhere(b) and z2(b) == 1 for b in B2)))

B1 count: 16
B2 count: 438
B1 z2 values: [0]
B2 z2 counts: {0: 294, 1: 144}
SU(2)-candidate B2 count: 16
SU(2)-candidate B2 z2 counts: {0: 8, 1: 8}
z2=1 B2 blocks not directly SU(2)-reducible: 136
z2=1 B2 blocks with equal C2 pair at some fixed K: 48


In [3]:
# Diagnostic only: which retained EBRs are themselves determinant-one two-band data?
print("EBRs that are determinant-one two-band blocks (original table columns):",
      [int(ebr_source_columns[j] + 1) for j, b in enumerate(EBR.T) if is_su2_candidate(b)])

# Audit the effect of reversing only the R coefficients.
# old P_1 - corrected P_1 = (-4,+4) on (R_3,R_4).
# For every two-band vector, R_3-R_4 is even, so the change is 0 mod 8.
# For every one-band vector, it is odd, so the old residue is 4 mod 8.
P_1_old = P_1.copy()
P_1_old[16:18] *= -1
assert np.all(B2 @ P_1_old % 8 == 0)
assert np.all(B1 @ P_1_old % 8 == 4)
print("R-sign correction preserves the entire two-band target set: True")
print("R-sign correction changes the one-band target set: True")

EBRs that are determinant-one two-band blocks (original table columns): []
R-sign correction preserves the entire two-band target set: True
R-sign correction changes the one-band target set: True


## Stabilized integer decomposition

The main check uses `su2_blocks_only = S` of shape `(20,16)` and `EBR`
of shape `(20,18)` to solve exactly the manuscript equation
`S @ u = b + EBR @ v` for all 438 two-band targets.
Every returned witness is checked for nonnegativity and exact integer equality.

Supplemental checks use (i) all 16 scalar U(1) generators together with S
for the one-band and two-band targets, and (ii) S alone for one-band targets.
Scalar blocks are zero-contribution U(1) blocks, not literal SU(1) blocks.
These supplemental checks are not assumptions of the manuscript's main check.

In [4]:
import highspy
from gu_cover_solver import build_highs_solver, solve_one_warm


def solve_targets(targets: np.ndarray, generators: np.ndarray, ebr: np.ndarray):
    """Solve generators @ U = target + ebr @ V for every target."""
    targets = np.asarray(targets, dtype=int)
    generators = np.asarray(generators, dtype=int)
    ebr = np.asarray(ebr, dtype=int)

    nU = generators.shape[1]
    A = np.ascontiguousarray(np.hstack([generators, -ebr]), dtype=np.float64)
    solver, row_idx = build_highs_solver(A, N_ROWS, A.shape[1])

    feasible = np.zeros(len(targets), dtype=bool)
    U_all = np.zeros((len(targets), nU), dtype=int)
    V_all = np.zeros((len(targets), ebr.shape[1]), dtype=int)
    failures: list[int] = []

    for i, target in enumerate(targets):
        ok, U, V = solve_one_warm(solver, row_idx, N_ROWS, nU, target)
        if ok:
            assert np.all(U >= 0) and np.all(V >= 0)
            assert np.all(generators @ U - ebr @ V == target)
            feasible[i] = True
            U_all[i] = U
            V_all[i] = V
        else:
            status = solver.getModelStatus()
            if status != highspy.HighsModelStatus.kInfeasible:
                raise RuntimeError(f"Target {i}: inconclusive solver status {status}")
            failures.append(i)
    return feasible, U_all, V_all, failures


t0 = time.time()
# Columns are generators. B1 blocks are scalar U(1) zero-contribution blocks, not SU(1) blocks.
allowed_blocks = np.vstack([B1, B2_su2]).T
all_targets = np.vstack([B1, B2])
feasible_allowed, U_allowed, V_allowed, failures_allowed = solve_targets(all_targets, allowed_blocks, EBR)

su2_blocks_only = B2_su2.T
feasible_b2_su2, U_b2_su2, V_b2_su2, failures_b2_su2 = solve_targets(B2, su2_blocks_only, EBR)
feasible_b1_su2, U_b1_su2, V_b1_su2, failures_b1_su2 = solve_targets(B1, su2_blocks_only, EBR)

print("allowed block matrix shape:", allowed_blocks.shape)
print("all 1D+2D targets feasible with scalar U(1)+SU(2) blocks:", bool(feasible_allowed.all()), f"({feasible_allowed.sum()}/{len(all_targets)})")
print("failures for scalar U(1)+SU(2) blocks:", failures_allowed)
print("all 2D targets feasible with SU(2) blocks only:", bool(feasible_b2_su2.all()), f"({feasible_b2_su2.sum()}/{len(B2)})")
print("failures for SU(2)-only 2D check:", failures_b2_su2)
print("all 1D targets feasible with SU(2) blocks only:", bool(feasible_b1_su2.all()), f"({feasible_b1_su2.sum()}/{len(B1)})")
print("failures for SU(2)-only 1D check:", failures_b1_su2)
print("max EBR columns added in scalar U(1)+SU(2) check:", int(V_allowed.sum(axis=1).max()))
print("max EBR columns added in SU(2)-only B2 check:", int(V_b2_su2.sum(axis=1).max()))
print("elapsed seconds:", round(time.time() - t0, 2))
assert feasible_allowed.all(), "Scalar U(1)+SU(2) decomposition has counterexamples"
assert feasible_b2_su2.all(), "Manuscript's SU(2)-only two-band claim has counterexamples"

allowed block matrix shape: (20, 32)
all 1D+2D targets feasible with scalar U(1)+SU(2) blocks: True (454/454)
failures for scalar U(1)+SU(2) blocks: []
all 2D targets feasible with SU(2) blocks only: True (438/438)
failures for SU(2)-only 2D check: []
all 1D targets feasible with SU(2) blocks only: True (16/16)
failures for SU(2)-only 1D check: []
max EBR columns added in scalar U(1)+SU(2) check: 2
max EBR columns added in SU(2)-only B2 check: 5
elapsed seconds: 17.93


## Counterexample report and the manuscript example

The main failure list concerns all 438 two-band targets with only S as generators.
The following cell uses the explicit target from Appendix `app:SU2_example`,
identified by its multiplicities rather than a hard-coded enumeration index.
It prints a solver witness; the next cell independently verifies the particular
three EBRs and three determinant-one summands printed in the manuscript.

In [5]:
if failures_allowed or failures_b2_su2:
    print("Counterexamples found.")
    if failures_allowed:
        print("scalar U(1)+SU(2) failed target indices:", failures_allowed)
    if failures_b2_su2:
        print("SU(2)-only B2 failed target indices:", failures_b2_su2)
else:
    print("No counterexamples found among all 16 one-band and 438 two-band blocks.")

# Appendix app:SU2_example, in the manuscript's unpermuted row convention.
example = np.zeros(N_ROWS, dtype=int)
example[[3, 5, 17, 19]] = 2
example[[8, 9, 14, 15]] = 1
example_matches = np.flatnonzero(np.all(B2 == example, axis=1))
assert len(example_matches) == 1
example_idx = int(example_matches[0])
assert z2(example) == 1 and not is_su2_candidate(example)
U = U_b2_su2[example_idx]
V = V_b2_su2[example_idx]
used_su2_local = np.nonzero(U)[0]
used_ebr = np.nonzero(V)[0]

print("\nRepresentative target: B2 index", example_idx, "(0-based), z2=", z2(example), "direct SU(2)=", is_su2_candidate(example))
print("target nonzero entries:", active_rep_labels(example))
print("added EBR columns (original table, 1-based, multiplicity):", [(int(ebr_source_columns[j] + 1), int(V[j])) for j in used_ebr])
print("SU(2) generator B2 indices used (0-based B2 index, multiplicity):", [(int(su2_block_ids[j]), int(U[j])) for j in used_su2_local])

lhs = su2_blocks_only @ U
rhs = example + EBR @ V
print("decomposition verified:", bool(np.array_equal(lhs, rhs)))
print("both sides nonzero entries:", active_rep_labels(lhs))

No counterexamples found among all 16 one-band and 438 two-band blocks.

Representative target: B2 index 26 (0-based), z2= 1 direct SU(2)= False
target nonzero entries: ['Gamma_8=2', 'A_6=2', 'M_5=1', 'M_6=1', 'Z_7=1', 'Z_8=1', 'R_4=2', 'X_4=2']
added EBR columns (original table, 1-based, multiplicity): [(15, 1), (22, 1), (36, 1)]
SU(2) generator B2 indices used (0-based B2 index, multiplicity): [(133, 1), (160, 1), (310, 1)]
decomposition verified: True
both sides nonzero entries: ['Gamma_5=1', 'Gamma_6=2', 'Gamma_7=1', 'Gamma_8=2', 'A_5=1', 'A_6=2', 'A_7=1', 'A_8=2', 'M_5=1', 'M_6=2', 'M_7=1', 'M_8=2', 'Z_5=2', 'Z_6=1', 'Z_7=2', 'Z_8=1', 'R_3=3', 'R_4=3', 'X_3=3', 'X_4=3']


In [6]:
# Verify the exact example printed in the manuscript, independently of the solver's choice.
# The listed EBRs are original MBANDREP columns 15, 22, 36 (1-based).
example_ebr_source = [15, 22, 36]
example_ebr_local = [int(np.flatnonzero(ebr_source_columns == j - 1)[0])
                     for j in example_ebr_source]
example_ebrs = EBR[:, example_ebr_local]

# Check the EBR eigenvalue lists through their representation multiplicities.
expected_ebrs = np.zeros((N_ROWS, 3), dtype=int)
expected_ebrs[[1, 4, 9, 12, 16, 18], 0] = 1
expected_ebrs[[2, 7, 11, 14, 16, 18], 1] = 1
expected_ebrs[[0, 1, 6, 7, 10, 11, 12, 13, 16, 17, 18, 19], 2] = 1
assert np.array_equal(example_ebrs, expected_ebrs)

# Table tab:det1-decomposition: small = exp(+/-i*pi/4), large = exp(+/-3i*pi/4).
pair_rows = {"small": [1, 3], "large": [0, 2]}
table_pairs = [
    ["small", "small", "large", "small"],  # Gamma,A,M,Z for Band 1&2
    ["small", "large", "small", "large"],  # Band 3&4
    ["large", "small", "small", "large"],  # Band 5&6
]
example_summands = np.zeros((N_ROWS, 3), dtype=int)
for j, choices in enumerate(table_pairs):
    for (_, idx), pair in zip(S4_BLOCKS, choices):
        example_summands[[idx[q] for q in pair_rows[pair]], j] = 1
    example_summands[16:20, j] = 1
    assert is_su2_candidate(example_summands[:, j])
    assert np.any(np.all(B2_su2 == example_summands[:, j], axis=1))
assert np.array_equal(example + example_ebrs.sum(axis=1), example_summands.sum(axis=1))
assert z2(example) == sum(z2(b) for b in example_summands.T) % 2
print("Manuscript's three EBR eigenvalue lists match source columns:", example_ebr_source)
print("Manuscript's three table columns belong to S: True")
print("Exact manuscript identity b + EBR_15 + EBR_22 + EBR_36 = sum(table columns): True")
print("z2 is preserved: True")

Manuscript's three EBR eigenvalue lists match source columns: [15, 22, 36]
Manuscript's three table columns belong to S: True
Exact manuscript identity b + EBR_15 + EBR_22 + EBR_36 = sum(table columns): True
z2 is preserved: True


## Rerun results and comparison with the manuscript

All code cells completed successfully with the corrected R sign.

| Check | Result |
| --- | --- |
| Admissible one-band / two-band vectors | 16 / 438 |
| Determinant-one two-band generators | 16 (8 with z2=0, 8 with z2=1) |
| Two-band targets stabilized using SU(2) generators only | 438 / 438 |
| One-band + two-band targets using scalar U(1) and SU(2) generators | 454 / 454 |
| One-band targets stabilized using SU(2) generators only | 16 / 16 |
| Explicit appendix example, including the three tabulated summands | Exact equality |

The two-band target set and the main 438/438 result are unchanged: reversing
the R coefficients changes z4S by 2*(m(R_3)-m(R_4)), which is a multiple of 4
when m(R_3)+m(R_4)=2. For one-band data it changes z4S by 2 modulo 4,
so the target set changes even though its size stays 16. The previous
notebook's 0/16 one-band SU(2)-only result and its negative conclusion no
longer apply to the corrected targets.

The main computation now uses exactly the manuscript's row convention,
indicator congruences, 18 double-valued EBRs, 16 determinant-one generators,
and equation `S @ u = b + EBR @ v` for all 438 two-band targets.
The code also checks the explicit appendix example by its multiplicities.

Differences from the previous notebook:
- The R coefficients of the unscaled z4S row are now (+1,-1), or (+2,-2) in P_1.
- Eighteen zero columns from single-valued EBRs are removed; original table
  column numbers are retained for traceability.
- Example vectors are checked explicitly instead of relying on stale indices.
- Supplemental one-band and scalar-generator checks are reported separately.

These are finite symmetry-data checks, not a construction of the smooth gauge
in the direct-reduction subsection. The Hilbert-basis calculation for arbitrary
rank belongs to the preceding block-decomposition section and is not rerun here.
In particular, numerical symmetry-data equality alone does not supply the
analytical justification for global smooth reblocking.

Minor manuscript correction still needed: replace "among them S columns" in
the paragraph giving the 438 targets with "among them 16 vectors". S denotes
the generator matrix, not a count.
